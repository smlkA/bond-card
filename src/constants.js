export const ReducerStates = {
  Regular: "regular",
  Saving: "saving",
};

export const Periods = {
  Week: "Week",
  Month: "Month",
  Quarter: "Quarter",
  Year: "Year",
};

export const yAxis = {
  Price: "price",
  Spread: "spread",
  Yield: "yield",
};

export const daysInPeriod = {
  Week: 7,
  Month: 30,
  Quarter: 90,
  Year: 365,
};

export const AXIS_DATE_FORMAT = "MM/dd";
