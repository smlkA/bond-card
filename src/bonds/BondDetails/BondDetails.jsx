import React from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { Typography, Spin, Radio, Divider, Select } from "antd";
import { LineChart, Line, CartesianGrid, XAxis, YAxis } from "recharts";
import { fetchBondInfo, setChartPeriod, setYAxis } from "../actions";
import {
  getBondIsin,
  getBondName,
  getLoadingState,
  getBondDescription,
  getBondData,
  getChartPeriod,
  getChartYAxis,
} from "../selector";
import { ReducerStates, Periods } from "../../constants";

const { Title, Paragraph } = Typography;
const { Option } = Select;

export class BondDetails extends React.Component {
  componentDidMount() {
    const { fetchBondInfo } = this.props;
    fetchBondInfo();
  }

  renderPeriodTabs() {
    const { period, setChartPeriod } = this.props;
    return (
      <Radio.Group
        value={period}
        buttonStyle="solid"
        onChange={(e) => setChartPeriod(e.target.value)}
      >
        <Radio.Button value={Periods.Week}>Week</Radio.Button>
        <Radio.Button value={Periods.Month}>Month</Radio.Button>
        <Radio.Button value={Periods.Quarter}>Quarter</Radio.Button>
        <Radio.Button value={Periods.Year}>Year</Radio.Button>
      </Radio.Group>
    );
  }

  renderSelect() {
    const { yAxis, setYAxis } = this.props;
    return (
      <Select
        value={yAxis}
        style={{ width: 120 }}
        onChange={(e) => {
          setYAxis(e);
        }}
      >
        <Option value="price">Price</Option>
        <Option value="spread">Spread</Option>
        <Option value="yield">Yield</Option>
      </Select>
    );
  }

  renderChart() {
    const { data, yAxis } = this.props;
    return (
      <LineChart width={600} height={300} data={data}>
        <Line dataKey={yAxis} stroke="#8884d8" />
        <CartesianGrid stroke="#ccc" />
        <XAxis dataKey="date" />
        <YAxis interval="preserveEnd" />
      </LineChart>
    );
  }

  render() {
    const { isin, bondName, description, loadingState } = this.props;

    return (
      <>
        {loadingState === ReducerStates.Saving ? (
          <Spin size="large" />
        ) : (
          <div>
            <Title>{bondName}</Title>
            <Paragraph>{isin}</Paragraph>
            <Paragraph>{description}</Paragraph>
            <Divider />
            {this.renderPeriodTabs()}
            {this.renderChart()}
            {this.renderSelect()}
          </div>
        )}
      </>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  isin: getBondIsin,
  bondName: getBondName,
  description: getBondDescription,
  loadingState: getLoadingState,
  data: getBondData,
  period: getChartPeriod,
  yAxis: getChartYAxis,
});

const mapDispatchToProps = {
  fetchBondInfo,
  setChartPeriod,
  setYAxis,
};

export default connect(mapStateToProps, mapDispatchToProps)(BondDetails);
