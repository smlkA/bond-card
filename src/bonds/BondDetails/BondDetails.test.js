import React from "react";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { BondDetails } from "./BondDetails";
import { Periods } from "../../constants";

const commonProps = {
  isin: "1",
  bondName: "Bond 1",
  period: Periods.Week,
  data: [],
  fetchBondInfo: Function.prototype,
};

describe("<BondDetails />", () => {
  afterEach(cleanup);

  it("should render BondDetails", () => {
    const wrapper = render(<BondDetails {...commonProps} />);
    expect(wrapper.asFragment()).toMatchSnapshot();
  });

  it("should call fetchBondInfo on componentDidMount", () => {
    const fetchBondInfo = jest.fn();

    render(<BondDetails {...commonProps} fetchBondInfo={fetchBondInfo} />);
    expect(fetchBondInfo).toHaveBeenCalled();
  });

  it("should call setChartPeriod on 'Month' tab click", () => {
    const setChartPeriod = jest.fn();

    const wrapper = render(
      <BondDetails {...commonProps} setChartPeriod={setChartPeriod} />
    );

    fireEvent.click(wrapper.getByLabelText("Month"));

    expect(setChartPeriod).toHaveBeenCalledWith(Periods.Month);
  });
});
