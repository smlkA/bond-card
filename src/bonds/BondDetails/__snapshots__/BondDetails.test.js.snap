// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BondDetails /> should render BondDetails 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="ant-typography"
    >
      Bond 1
    </h1>
    <div
      class="ant-typography"
    >
      1
    </div>
    <div
      class="ant-typography"
    />
    <div
      class="ant-divider ant-divider-horizontal"
      role="separator"
    />
    <div
      class="ant-radio-group ant-radio-group-solid"
    >
      <label
        class="ant-radio-button-wrapper ant-radio-button-wrapper-checked"
      >
        <span
          class="ant-radio-button ant-radio-button-checked"
        >
          <input
            checked=""
            class="ant-radio-button-input"
            type="radio"
            value="Week"
          />
          <span
            class="ant-radio-button-inner"
          />
        </span>
        <span>
          Week
        </span>
      </label>
      <label
        class="ant-radio-button-wrapper"
      >
        <span
          class="ant-radio-button"
        >
          <input
            class="ant-radio-button-input"
            type="radio"
            value="Month"
          />
          <span
            class="ant-radio-button-inner"
          />
        </span>
        <span>
          Month
        </span>
      </label>
      <label
        class="ant-radio-button-wrapper"
      >
        <span
          class="ant-radio-button"
        >
          <input
            class="ant-radio-button-input"
            type="radio"
            value="Quarter"
          />
          <span
            class="ant-radio-button-inner"
          />
        </span>
        <span>
          Quarter
        </span>
      </label>
      <label
        class="ant-radio-button-wrapper"
      >
        <span
          class="ant-radio-button"
        >
          <input
            class="ant-radio-button-input"
            type="radio"
            value="Year"
          />
          <span
            class="ant-radio-button-inner"
          />
        </span>
        <span>
          Year
        </span>
      </label>
    </div>
    <div
      class="recharts-wrapper"
      style="position: relative; cursor: default; width: 600px; height: 300px;"
    >
      <svg
        class="recharts-surface"
        height="300"
        version="1.1"
        viewBox="0 0 600 300"
        width="600"
      >
        <defs>
          <clippath
            id="recharts2-clip"
          >
            <rect
              height="260"
              width="530"
              x="65"
              y="5"
            />
          </clippath>
        </defs>
        <g
          class="recharts-cartesian-grid"
        >
          <g
            class="recharts-cartesian-grid-horizontal"
          >
            <line
              fill="none"
              height="260"
              stroke="#ccc"
              width="530"
              x="65"
              x1="65"
              x2="595"
              y="5"
              y1="5"
              y2="5"
            />
            <line
              fill="none"
              height="260"
              stroke="#ccc"
              width="530"
              x="65"
              x1="65"
              x2="595"
              y="5"
              y1="265"
              y2="265"
            />
          </g>
          <g
            class="recharts-cartesian-grid-vertical"
          >
            <line
              fill="none"
              height="260"
              stroke="#ccc"
              width="530"
              x="65"
              x1="65"
              x2="65"
              y="5"
              y1="5"
              y2="265"
            />
            <line
              fill="none"
              height="260"
              stroke="#ccc"
              width="530"
              x="65"
              x1="595"
              x2="595"
              y="5"
              y1="5"
              y2="265"
            />
          </g>
        </g>
        <g
          class="recharts-layer recharts-cartesian-axis recharts-yAxis yAxis"
        >
          <line
            class="recharts-cartesian-axis-line"
            fill="none"
            height="260"
            stroke="#666"
            width="60"
            x="5"
            x1="65"
            x2="65"
            y="5"
            y1="5"
            y2="265"
          />
          <g
            class="recharts-cartesian-axis-ticks"
          />
        </g>
      </svg>
    </div>
    <div
      class="ant-select ant-select-single ant-select-show-arrow"
      style="width: 120px;"
    >
      <div
        class="ant-select-selector"
      >
        <span
          class="ant-select-selection-search"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="ant-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            value=""
          />
        </span>
        <span
          class="ant-select-selection-placeholder"
        />
      </div>
      <span
        aria-hidden="true"
        class="ant-select-arrow"
        style="user-select: none;"
        unselectable="on"
      >
        <span
          aria-label="down"
          class="anticon anticon-down"
          role="img"
        >
          <svg
            aria-hidden="true"
            class=""
            data-icon="down"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
            />
          </svg>
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
